### ISSUE TITLE
🐛 [Transmission OpenVPN] Mullvad provider does not work without ipv6 enabled

### ISSUE THREAD

**Issue by AndrewVos** [alexbelgium/hassio-addons#494]
The latest version of "Transmission OpenVPN" can't work with Mullvad because it requires specific settings when launching the container:

https://haugene.github.io/docker-transmission-openvpn/provider-specific/#mullvad_ovpn

I'm not quite sure how to resolve this - does anyone know how to pass parameters to the container?

Great work on this addon by the way!! Absolutely amazing stuff <3

**Comment by alexbelgium** [comment]
Hi, it's not possible to use custom docker arguments. The only way is through standardized HA config.json that don't include ipv6 options. However, ipv6 should be in theory enables by default in HA

**Comment by AndrewVos** [comment]
That's interesting. I wonder if anyone else has a way of getting mullvad working. I assume someone here has got it working before if it's in the provider list

**Comment by alexbelgium** [comment]
Actually I just put the full list available from the upstream container (https://github.com/haugene/docker-transmission-openvpn) but didn't test them... And given it uses the connection from the docker installation i suppose it would depend on it is configured more than how the addon is built

**Comment by AndrewVos** [comment]
Oh cool! I think I'll just try a different VPN. Thanks for your help!

**Comment by jaamulberry** [comment]
If you use the qbittorrent add-on you can disable ipv6 and use Mullvad. Just need to add some lines to the ovpn file: 

pull-filter ignore "route-ipv6"
pull-filter ignore "ifconfig-ipv6"

**Comment by cramertj** [comment]
Hey! I'm having this same issue-- it seems like the answer is that it isn't possible to support?

> Hi, it's not possible to use custom docker arguments. The only way is through standardized HA config.json that don't include ipv6 options. However, ipv6 should be in theory enables by default in HA

Am I understanding correctly that this is a restriction of Home Assistant OS itself? If so, I'll file a bug there and see if we can get an option added to configure the appropriate ipv6 setting.

**Comment by alexbelgium** [comment]
Well from what I read above modifying the mullvad ovpn to disable IPv6 could work ? That I could do

**Comment by alexbelgium** [comment]
(code not tested)

**Comment by Blogshot** [comment]
I've got a solution! See below for my way to a solution, ther may be better ways to do it.

I hope this helps and that other will be able to fix this as well.

Steps:
1. `docker exec -ti addon_db21ed7f_transmission_openvpn bash` to get into the container
2. Create `/etc/openvpn/mullvad/configure-openvpn-sh` with the following contents
```
echo "Removing IPv6 from mullvad"

echo "pull-filter ignore \"route-ipv6\"" | tee -a "/etc/openvpn/mullvad/"*.ovpn
echo "pull-filter ignore \"ifconfig-ipv6\"" | tee -a "/etc/openvpn/mullvad/"*.ovpn
echo "pull-filter ignore \"dhcp-option DNS6\"" | tee -a "/etc/openvpn/mullvad/"*.ovpn
```
3. Restart the container and check the logs
4. Profit  

After restarting the container, it works:
```
2022-11-29 10:34:10 TUN/TAP device tun0 opened
2022-11-29 10:34:10 net_iface_mtu_set: mtu 1500 for tun0
2022-11-29 10:34:10 net_iface_up: set tun0 up
2022-11-29 10:34:10 net_addr_v4_add: 10.9.0.34/16 dev tun0
2022-11-29 10:34:10 net_route_v4_add: 185.65.134.137/32 via 172.30.32.1 dev [NULL] table 0 metric -1
2022-11-29 10:34:10 net_route_v4_add: 0.0.0.0/1 via 10.9.0.1 dev [NULL] table 0 metric -1
2022-11-29 10:34:10 net_route_v4_add: 128.0.0.0/1 via 10.9.0.1 dev [NULL] table 0 metric -1
Up script executed with device=tun0 ifconfig_local=10.9.0.34
Updating TRANSMISSION_BIND_ADDRESS_IPV4 to the ip of tun0 : 10.9.0.34
...
STARTING TRANSMISSION
Transmission startup script complete.
2022-11-29 10:34:10 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
2022-11-29 10:34:10 Initialization Sequence Completed
[10:34:11] INFO: Ingress ready
```

Research: The issue is in `/etc/openvpn/start.sh`:

This whole block does not resolve to either block, which results in `CHOSEN_OPENVPN_CONFIG` to not be set.
```
echo "Using OpenVPN provider: ${VPN_PROVIDER^^}"
if [[ "${VPN_PROVIDER}" == "custom" ]]; then
  if [[ -x $VPN_PROVIDER_HOME/default.ovpn ]]; then
    echo "Found default config: $VPN_PROVIDER_HOME/default.ovpn"
    CHOSEN_OPENVPN_CONFIG=$VPN_PROVIDER_HOME/default.ovpn
  fi
elif [[ -n $OPENVPN_CONFIG_URL ]]; then
  echo "Found URL to single OpenVPN config, will download and use it."
  CHOSEN_OPENVPN_CONFIG=$VPN_PROVIDER_HOME/downloaded_config.ovpn
  curl -o "$CHOSEN_OPENVPN_CONFIG" -sSL "$OPENVPN_CONFIG_URL"
fi
```

As a fallback, it will set to `auto`.
```
if [[ -z ${CHOSEN_OPENVPN_CONFIG} ]]; then

  # Support pulling configs from external config sources
  VPN_CONFIG_SOURCE="${VPN_CONFIG_SOURCE:-auto}"
  VPN_CONFIG_SOURCE="${VPN_CONFIG_SOURCE,,}" # to lowercase

  echo "Running with VPN_CONFIG_SOURCE ${VPN_CONFIG_SOURCE}"
```

`$VPN_PROVIDER_HOME/configure-openvpn.sh` does not exist, so we land in the last `else` block. This will set `VPN_CONFIG_SOURCE=external` and trigger `./etc/openvpn/fetch-external-configs.sh`, which then downloads the configs again and overwrites our changes.
```
  if [[ "${VPN_CONFIG_SOURCE}" == "auto" ]]; then
    if [[ -x $VPN_PROVIDER_HOME/configure-openvpn.sh ]]; then
      echo "Provider ${VPN_PROVIDER^^} has a bundled setup script. Defaulting to internal config"
      VPN_CONFIG_SOURCE=internal
    elif [[ "${VPN_PROVIDER}" == "custom" ]]; then
      echo "CUSTOM provider specified but not using default.ovpn, will try to find a valid config mounted to $VPN_PROVIDER_HOME"
      VPN_CONFIG_SOURCE=custom
    else
      echo "No bundled config script found for ${VPN_PROVIDER^^}. Defaulting to external config"
      VPN_CONFIG_SOURCE=external
    fi
  fi

  if [[ "${VPN_CONFIG_SOURCE}" == "external" ]] && [[ "${VPN_PROVIDER}" != "custom" ]]; then
    # shellcheck source=openvpn/fetch-external-configs.sh
    ./etc/openvpn/fetch-external-configs.sh
  fi
```

To summarize:
1. We change the necessary settings in the opvn-files
6. The startup-script ignores the local config files and overwrites them
7. The changes are lost
8. One possible solution is to provide `$VPN_PROVIDER_HOME/configure-openvpn.sh`, which will skip the download from external.

**Comment by alexbelgium** [comment]
Thanks! Would it work if I modify on the fly using a script at startup of the addon ? so that it uses the default mullvad ovpn config that is downloaded from the repo, then adds the lines you mentioned?

**Comment by Blogshot** [comment]
The change would need to happen somewhere after [line 98](https://github.com/haugene/docker-transmission-openvpn/blob/b33d0fe4c938259a0d4eb844e55468f387456121/openvpn/start.sh#L98) of haugene's script. The only way I've found is to place `configure-openvpn.sh` because the script in it's original form calls it natively.

To insert a custom script you'd need to manipulate that file at startup, but it would work.

Another note: You commits did actually break the Mullvad functionality, see my comment on https://github.com/alexbelgium/hassio-addons/commit/77dc2d933e69a2b49edad33a321d9befc166dd77.

**Comment by alexbelgium** [comment]
Thanks very much for your research and clear solution. I'll follow your recommendation in the last message

**Comment by alexbelgium** [comment]
Actually I had misread your first solution... Just adding /etc/openvpn/mullvad/configure-openvpn.sh should then work? Then I've added it to the base image. Actually if this when it is indeed more elegant the using the proposed sed to modify the start script. Sorry

**Comment by Blogshot** [comment]
Hey there!

I just updated to v4.3.2-6 and the script does not seem to be integrated correctly. I can see it under `/usr/share/hassio/addons/git/db21ed7f/transmission_openvpn/rootfs/etc/openvpn/mullvad/configure-openvpn.sh`, but in the container itself exists no `/etc/openvpn/mullvad/configure-openvpn.sh`.

The script also has no executable flag it it's current state, which will lead to the same behaviour because [`if [[ -x $VPN_PROVIDER_HOME/configure-openvpn.sh ]]; then`](https://github.com/haugene/docker-transmission-openvpn/blob/b33d0fe4c938259a0d4eb844e55468f387456121/openvpn/start.sh#L82) specifically checks if the file is executable.

I just noticed this while searching for the issue a few minutes ago, so please forgive me for not mentioning that flag in detail. 😅 

After `chmod +x`'ing the script, I tried to copy the script manually into the container, but the container seems to clean itself at restart so the script gets deleted again. 
However, I did succeed using a workaround: In the first few seconds after container start, I issue a quick `docker cp ./configure-openvpn.sh addon_db21ed7f_transmission_openvpn:/etc/openvpn/mullvad/`, which luckily happens after the purge. With that, the script was correctly in place and the addon connection works flawlessy, as far as I can tell.

I don't know how your containers exactly work, but I guess the new file must somehow be added to a list of files to be copied at some point? Or maybe something in the docker build pipeline?

**Comment by alexbelgium** [comment]
Mmh normally the file is copied there during build through the `COPY rootfs/ /` command...
I'll add the `chmod +x` to the run script to see if it helps. I'll try with my free VPN if the file exists

**Comment by alexbelgium** [comment]
Well the file exists when the container starts. Ok have a script that executes /config/transmission-openvpn.sh if the file exists. I've added a line to echo if the file /etc/openvpn/mullvad/configure-openvpn.sh exists and it does. Could it be that the Haugen script purges the /etc/openvpn folder ?

**Comment by Blogshot** [comment]
Ok, I think I've found out that the problem was on my side.

Because the script was not executable on my first attempt, Haugen's script downloaded (and overwrote) the files, removing the script alongside.

The executable flag should fix that behaviour, I will test as soon as HA reports about the update.

**Comment by Blogshot** [comment]
I just did a clean uninstall and reinstall and the script is correctly present, great!

Now, another issue unfolded with this.
- Haugen's script only downloads the default openvpn files (which are needed) when there is no script inside `/etc/openvpn/mullvad`.
- Because we created that script, Haugen's script won't download the needed default files which we wanted to modify in the first place. 

We basically created a little chicken-egg problem where we either have the default files or the script, but never both. 😢 
I don't really know how to tackle this in an elegant way.

It could be solved by providing the default files within your image from the get-go, which complicates the docker build a bit since you need to download and unzip the files from Haugen's repository.
Alternatively you could just manually copy them into your own repository, but this would create problems in the future if Haugen updates them at some point.

The relevant parts of that download (taken and simplified from Haugen's `/etc/openvpn/fetch-external-configs.sh`:
```
  GITHUB_CONFIG_BUNDLE_URL="https://github.com/haugene/vpn-configs-contrib/archive/main.zip"

  # Create a temporary file and download bundle to it
  config_repo_temp_zip_file=$(mktemp)
  echo "Downloading configs from ${GITHUB_CONFIG_BUNDLE_URL} into ${config_repo_temp_zip_file}"
  curl -sSL --fail -o "${config_repo_temp_zip_file}" "${GITHUB_CONFIG_BUNDLE_URL}"

  # Create a temporary folder and extract configs there
  config_repo_temp_dir=$(mktemp -d)
  echo "Extracting configs to ${config_repo_temp_dir}"
  unzip -q "${config_repo_temp_zip_file}" -d "${config_repo_temp_dir}"

  # Find the specified provider folder. Should be under <tmpDir>/<some-root-folder>/openvpn/<provider>
  provider_configs=$(find "${config_repo_temp_dir}"/*/openvpn -type d -name "${VPN_PROVIDER}")
  if [[ -z "${provider_configs}" ]]; then
    echo "ERROR: Could not find any configs for provider ${VPN_PROVIDER^^} in downloaded configs"
    exit 1
  fi

  # Replace current provider home folder with the downloaded directory
  echo "Found configs for ${VPN_PROVIDER^^} in ${provider_configs}, will replace current content in ${VPN_PROVIDER_HOME}"
  find "${provider_configs}" -type f -exec cp '{}' ${VPN_PROVIDER_HOME}/ \;
```

Functionally, this is the same as the download if no `configure-openvpn.sh` script would exist _except_ it won't remove existing files. With this, we can create the configuration-script and also get the latest `*.ovpn` files.

Is it possible to try this approach?  I really appreciate your fast reactions up to this point. 👍

**Comment by alexbelgium** [comment]
Ahah, always a new challenge. This approach with sed that you proposed should work though at it modifies the files after having downloaded them
```
To modify the original script:
I think something like sed -i '$i/opt/modify-mullvad.sh' /etc/openvpn/start.sh at startup should work. This sed command inserts a call to /opt/modify-mullvad.sh into the second last line of /etc/openvpn/start.sh, right before the OpenVPN executable gets called.
```

**Comment by Blogshot** [comment]
Yes, that should be sufficient I think. You'd have to move the configuration script back to e.g. `/opt/` and then the sed in 99-run.sh.

Basically you could revert to https://github.com/alexbelgium/hassio-addons/commit/62c6349560e06119676405cbf80cde1647cdcb9d and add the call to `sed`.

**Comment by Blogshot** [comment]
Hey there!

I've updated the container and everything seems fine! 

It correctly downloads the default scripts and then adds the rules regarding IPv6 to the `*.ovpn` files. I can see that OpenVPN correctly recognizes the filters and removes everything IPv6 related. ❤️ 

I tested the connection using https://ipleak.net/ and it reports my remote IP, so everything seems to work fine. 

I'll finally be able to switch from qBitTorrent to Transmission, which seems like a more stable and reliable option regarding IP leakage.

For more information, here is a censored excerpt from my logs:
```
Running with VPN_CONFIG_SOURCE auto
No bundled config script found for MULLVAD. Defaulting to external config
Downloading configs from https://github.com/haugene/vpn-configs-contrib/archive/main.zip into /tmp/tmp.mzSV7lzWHi
[...]
Removing IPv6 from mullvad
pull-filter ignore "route-ipv6"
pull-filter ignore "ifconfig-ipv6"
pull-filter ignore "dhcp-option DNS6"
[...]
2022-11-30 14:30:58 [yy-yyy-yyy.mullvad.net] Peer Connection Initiated with [AF_INET]xxx.xxx.xxx.xxx:1195
[...]
2022-11-30 14:31:00 Pushed option removed by filter: 'route-ipv6 0000::/2'
2022-11-30 14:31:00 Pushed option removed by filter: 'route-ipv6 4000::/2'
2022-11-30 14:31:00 Pushed option removed by filter: 'route-ipv6 8000::/2'
2022-11-30 14:31:00 Pushed option removed by filter: 'route-ipv6 C000::/2'
2022-11-30 14:31:00 Pushed option removed by filter: 'ifconfig-ipv6 zzz/64 zzz::'
[...]
Up script executed with device=tun0 ifconfig_local=xxx.xxx.xxx.xxx
Updating TRANSMISSION_BIND_ADDRESS_IPV4 to the ip of tun0 : xxx.xxx.xxx.xxx
[...]
```
